### ISSUE TITLE
Possible memory leak on production

### ISSUE THREAD

**Issue by danilobreda** [TykTechnologies/tyk-pump#40]
I had this memory problem today, since i'm using dynatrace as a monitoring service, i can have the timeline of whats happened.

> Jul 13 18:42:52 ip-172-31-11-143 kernel: [1135762.108931] Out of memory: Kill process 27863 (tyk-pump) score 507 or sacrifice child
> Jul 13 18:42:52 ip-172-31-11-143 kernel: [1135762.115685] Killed process 27863 (tyk-pump) total-vm:701084kB, anon-rss:530936kB, file-rss:0kB
> Jul 13 18:42:52 ip-172-31-11-143 kernel: [1135762.209707] init: tyk-pump main process (27863) killed by KILL signal
> Jul 13 18:42:52 ip-172-31-11-143 kernel: [1135762.209719] init: tyk-pump main process ended, respawning

![memoryleak](https://user-images.githubusercontent.com/14843499/28190237-4a7e0afe-6800-11e7-8dfd-5ff19a6fb41a.png)

**Comment by mvdan** [comment]
What version of tyk-pump is this?

**Comment by danilobreda** [comment]
this is when i start the pump...
Version: v0.4.1.1
> root@ip-172-31-11-143:/opt/tyk-pump# ./tyk-pump
> [Jul 14 12:49:25]  INFO main: ## Tyk Analytics Pump, v0.4.1.1 ##
> [Jul 14 12:49:25]  INFO mongo-pump: -- No max batch size set, defaulting to 10MB
> [Jul 14 12:49:25]  INFO mongo-pump: -- No max document size set, defaulting to 10MB
> [Jul 14 12:49:25]  INFO main: Init Pump: MongoDB Pump
> [Jul 14 12:49:25]  INFO main: Init Pump: MongoDB Aggregate Pump
> [Jul 14 12:49:25]  INFO mongo-pump: -- No max batch size set, defaulting to 10MB
> [Jul 14 12:49:25]  INFO mongo-pump: -- No max document size set, defaulting to 10MB
> [Jul 14 12:49:25]  INFO main: Init Uptime Pump: MongoDB Pump
> [Jul 14 12:49:25]  INFO main: Starting purge loop @10(s)
> [Jul 14 12:49:35]  WARN redis: Connection dropped, connecting..
> [Jul 14 12:49:35]  WARN redis: Connection dropped, connecting..
> [Jul 14 12:51:25]  INFO mongo-pump-aggregate: Writing 2 records
> [Jul 14 12:51:25]  INFO mongo-pump: Purging 2 records

no problems here, it seems that after the leak(i rebooted the server) its ok..

config:

```
{
    "analytics_storage_type": "redis",
    "analytics_storage_config": {
        "type": "redis",
        "host": "localhost",
        "port": 6379,
        "hosts": null,
        "username": "",
        "password": "",
        "database": 0,
        "optimisation_max_idle": 100,
        "optimisation_max_active": 0,
        "enable_cluster": false
    },
    "purge_delay": 10,
    "pumps": {
        "mongo": {
            "name": "mongo",
            "meta": {
                "collection_name": "tyk_analytics",
                "mongo_url": "mongodb://127.0.0.1/tyk_analytics"
            }
        },
        "mongo-pump-aggregate": {
            "name": "mongo-pump-aggregate",
            "meta": {
                "mongo_url": "mongodb://127.0.0.1/tyk_analytics",
                "use_mixed_collection": true
            }
        }
    },
    "uptime_pump_config": {
        "collection_name": "tyk_uptime_analytics",
        "mongo_url": "mongodb://127.0.0.1/tyk_analytics"
    },
    "dont_purge_uptime_data": false
}

```

i have a expire.txt that i configured on mongodb to drop old data too... (the problem can be here?)

```
 {
    "allowance": 10000,
    "rate": 10000,
    "per": 0,
    "expires": 0,
    "quota_max": -1,
    "quota_renews": 1406121006,
    "quota_remaining": 100000,
    "quota_renewal_rate": 100000,
    "org_id": "58754c722658ae0d071bc44b",
    "data_expires": 86400
    }
```

**Comment by danilobreda** [comment]
@mvdan is going very weird... i have a month image now...

![monthpng](https://user-images.githubusercontent.com/14843499/28643585-af93d7d0-722c-11e7-97bf-a86bbdd2faf7.png)

Reinstall? disable it temporary? what can i do?

**Comment by lonelycode** [comment]
I'd suggest upgrade to the latest pump (0.4.2), there was a memleak that has been fixed since. Also, decrease the purge time to like 5s, means smaller batch writes to Mongo and smaller processing packages.

**Comment by danilobreda** [comment]
Problem resolved after upgrading to latest pump 0.4.2. Purge time decreased. Now pump is using 20MB to 50MB of memory on certain periods. Thanks 👍
